src/targets/registry.ts

```diff
@@ -408,20 +445,76 @@ export class RegistryTarget extends BaseTarget {
         this.logger.warn(
           `No files found that match "${onlyIfPresentPattern.toString()}", skipping the target.`
         );
+        this.published = true;
         return undefined;
       }
     }
 
-    const remote = this.registryConfig.registryRemote;
-    const username = await getAuthUsername(this.github);
-    remote.setAuth(username, getGithubApiToken());
-
-    await withTempDir(
-      directory =>
-        this.pushVersionToRegistry(directory, remote, version, revision),
-      true,
-      'craft-release-registry-'
+    const startBatch = RegistryTarget.lock.isLocked();
+    const result = RegistryTarget.lock.runExclusive(() =>
+      this.doPublish(version, revision)
     );
-    this.logger.info('Release registry updated');
+    if (startBatch) {
+      const batch = [];
+      for (const instance of RegistryTarget.instances) {
+        batch.push(
+          instance === this ? result : instance.publish(version, revision)
+        );
+      }
+      return Promise.all(batch);
+    }
+    return result;
+  }
+
+  private async doPublish(version: string, revision: string) {
+    if (!RegistryTarget.localRepo) {
+      await withTempDir(
+        async dir => {
+          RegistryTarget.localRepo = {
+            dir,
+            git: await this.cloneRegistry(dir),
+          };
+        },
+        // We will clean the directory after pushing
+        false,
+        'craft-release-registry-'
+      );
+    }
+    let localRepo: LocalRegistry;
+    if (RegistryTarget.localRepo) {
+      localRepo = RegistryTarget.localRepo;
+    } else {
+      // XXX(BYK): This should NEVER happen
+      throw new Error(
+        `Local registry missing, it should have been cloned at this stage!`
+      );
+    }
+    this.commitVersionToRegistry(localRepo, version, revision);
+
+    if (this.isLastToPublish()) {
+      // Push!
+      if (!isDryRun()) {
+        this.logger.info(`Pushing the changes...`);
+        // Ensure we are still up to date with upstream
+        await withRetry(() =>
+          localRepo.git
+            .pull('origin', 'master', ['--rebase'])
+            .push('origin', 'master')
```

Comment:
Rebasing minimizes but doesn't eliminate the race condition with other releases that might be going on. Probably low enough traffic that this is fine for now, might matter some day.

Reply:
Note the `withRetry` that's wrapping this so we try up to 3 times. See #232 for more context.